File: android/src/main/java/com/margelo/nitro/qusaieilouti99/callmanager/CallManager.kt
package com.margelo.nitro.qusaieilouti99.callmanager

import android.util.Log
import com.facebook.proguard.annotations.DoNotStrip
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

@DoNotStrip
class CallManager : HybridCallManagerSpec() {

    private val TAG = "CallManager"
    private val mainScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    private fun ensureInitialized() {
        if (!CallEngine.isInitialized()) {
            Log.e(TAG, "CallEngine not initialized! This should not happen if Application.onCreate() was called properly.")
            throw IllegalStateException(
                "CallEngine must be initialized in Application.onCreate(). " +
                    "Make sure MainApplication.onCreate() calls CallEngine.initialize(this) before any native calls."
            )
        }
    }

    override fun endCall(callId: String): Unit {
        Log.d(TAG, "endCall requested for callId: $callId")
        ensureInitialized()
        CallEngine.endCall(callId)
    }

    override fun endAllCalls(): Unit {
        Log.d(TAG, "endAllCalls requested")
        ensureInitialized()
        CallEngine.endAllCalls()
    }

    override fun silenceRingtone(): Unit {
        Log.d(TAG, "silenceRingtone requested.")
        ensureInitialized()
        CallEngine.stopRingtone()
    }

    override fun getAudioDevices(): AudioRoutesInfo {
        Log.d(TAG, "getAudioDevices requested.")
        ensureInitialized()
        return CallEngine.getAudioDevices()
    }

    override fun setAudioRoute(route: String): Unit {
        Log.d(TAG, "setAudioRoute requested for route: $route")
        ensureInitialized()
        CallEngine.setAudioRoute(route)
    }

    override fun keepScreenAwake(keepAwake: Boolean): Unit {
        Log.d(TAG, "keepScreenAwake requested: $keepAwake")
        ensureInitialized()
        CallEngine.setIdleTimerDisabled(keepAwake)
    }

    override fun addCallListener(listener: (event: CallEventType, payload: String) -> Unit): Unit {
        Log.d(TAG, "addCallListener called")
        ensureInitialized()

        val wrapped: (CallEventType, String) -> Unit = { event, payload ->
            mainScope.launch {
                try {
                    listener(event, payload)
                } catch (e: Exception) {
                    Log.e(TAG, "Error executing JS call listener", e)
                }
            }
        }

        CallEngine.setEventHandler(wrapped)
    }

    override fun removeCallListener(): Unit {
        Log.d(TAG, "removeCallListener called")
        ensureInitialized()
        CallEngine.setEventHandler(null)
    }

    override fun registerVoIPTokenListener(listener: (payload: String) -> Unit): Unit {
        Log.d(TAG, "registerVoIPTokenListener called")

    }

    override fun removeVoipTokenListener(): Unit {
        Log.d(TAG, "removeVoipTokenListener called")

    }

    override fun startOutgoingCall(callId: String, callType: String, targetName: String, metadata: String?): Unit {
        Log.d(TAG, "startOutgoingCall requested: callId=$callId, callType=$callType, targetName=$targetName")
        ensureInitialized()
        CallEngine.startOutgoingCall(callId, callType, targetName, metadata)
    }

    override fun startCall(callId: String, callType: String, targetName: String, metadata: String?): Unit {
        Log.d(TAG, "startCall requested: callId=$callId, callType=$callType, targetName=$targetName")
        ensureInitialized()
        CallEngine.startCall(callId, callType, targetName, metadata)
    }

    override fun callAnswered(callId: String): Unit {
        Log.d(TAG, "callAnswered (from JS) requested for callId: $callId")
        ensureInitialized()
        CallEngine.answerCall(callId, isLocalAnswer = false)
    }

    override fun setOnHold(callId: String, onHold: Boolean): Unit {
        Log.d(TAG, "setOnHold requested for callId: $callId, onHold: $onHold")
        ensureInitialized()
        CallEngine.setOnHold(callId, onHold)
    }

    override fun setMuted(callId: String, muted: Boolean): Unit {
        Log.d(TAG, "setMuted requested for callId: $callId, muted: $muted")
        ensureInitialized()
        CallEngine.setMuted(callId, muted)
    }

    override fun updateDisplayCallInformation(callId: String, callerName: String): Unit {
        // no-op
    }

    override fun reportIncomingCall(
        callId: String,
        callType: String,
        targetName: String,
        metadata: String?,
        token: String?,
        rejectEndpoint: String?
    ) {
        Log.d(TAG, "reportIncomingCall requested: callId=$callId, callType=$callType, targetName=$targetName")
        ensureInitialized()
        CallEngine.reportIncomingCall(
            requireNotNull(CallEngine.getContext()) { "CallEngine must be initialized with context" },
            callId,
            callType,
            targetName,
            null,
            metadata,
            token,
            rejectEndpoint
        )
    }

    override fun hasActiveCall(): Boolean {
        Log.d(TAG, "hasActiveCall (from JS)")
        ensureInitialized()
        return CallEngine.isCallActive()
    }

    override fun requestOverlayPermissionAndroid(): Boolean {
        Log.d(TAG, "requestOverlayPermissionAndroid requested (launch settings)")
        ensureInitialized()
        val ctx = requireNotNull(CallEngine.getContext()) { "CallEngine must be initialized with context" }
        return CallEngine.launchOverlayPermissionSettings(ctx)
    }

    override fun hasOverlayPermissionAndroid(): Boolean {
        Log.d(TAG, "hasOverlayPermissionAndroid requested (check only)")
        ensureInitialized()
        val ctx = requireNotNull(CallEngine.getContext()) { "CallEngine must be initialized with context" }
        return CallEngine.checkOverlayPermissionGranted(ctx)
    }
}

File: ios/CallManager.swift
import Foundation
import NitroModules
import OSLog
import UIKit

/// The public-facing bridge class that implements the `HybridCallManagerSpec` protocol.
/// Its responsibility is to translate calls from the JavaScript layer into commands
/// for the `CallEngine` singleton. It acts as a thin, safe interface to the native call logic.
public class CallManager: HybridCallManagerSpec {
    private let logger = Logger(
        subsystem: "com.qusaieilouti99.callmanager",
        category: "CallManager"
    )

    public override init() {
        super.init()
        logger.info("🎯 CallManager hybrid module init")
        CallEngine.shared.initialize()
        logger.info("🎯 ✅ CallManager initialized")
    }

    public func endCall(callId: String) throws {
        logger.info("🎯 endCall ▶ js → native: \(callId)")
        CallEngine.shared.endCall(callId: callId)
    }

    public func endAllCalls() throws {
        logger.info("🎯 endAllCalls ▶ js → native")
        CallEngine.shared.endAllCalls()
    }

    public func silenceRingtone() throws {
        logger.info("🎯 silenceRingtone ▶ js → native (handled by CallKit)")
    }

    public func getAudioDevices() throws -> AudioRoutesInfo {
        logger.info("🎯 getAudioDevices ▶ js → native")
        return CallEngine.shared.getAudioDevices()
    }

    public func setAudioRoute(route: String) throws {
        logger.info("🎯 setAudioRoute ▶ js → native: \(route)")
        CallEngine.shared.setAudioRoute(route: route)
    }

    public func keepScreenAwake(keepAwake: Bool) throws {
        logger.info("🎯 keepScreenAwake ▶ js → native: \(keepAwake)")
        CallEngine.shared.setIdleTimerDisabled(shouldDisable: keepAwake)
    }

    public func addListener(listener: @escaping (CallEventType, String) -> Void) throws
        -> () -> Void
    {
        logger.info("🎯 addListener ▶ js → native")
        CallEngine.shared.setEventHandler { event, payload in
            self.logger.debug("🎯 event \(event.stringValue), payload.len=\(payload.count)")
            listener(event, payload)
        }
        // Return a closure that will be called by the JS layer to unsubscribe.
        return {
            self.logger.info("🎯 removeListener ▶ js → native")
            CallEngine.shared.setEventHandler(nil)
        }
    }

    public func startOutgoingCall(
        callId: String,
        callType: String,
        targetName: String,
        metadata: String?
    ) throws {
        logger.info("🎯 startOutgoingCall ▶ js → native: \(callId), type=\(callType)")
        CallEngine.shared.startOutgoingCall(
            callId: callId,
            callType: callType,
            targetName: targetName,
            metadata: metadata
        )
    }

    public func reportIncomingCall(
        callId: String,
        callType: String,
        targetName: String,
        metadata: String?,
        token: String?,
        rejectEndpoint: String?
    ) throws {
        logger.info("🎯 reportIncomingCall ▶ js → native: \(callId), type=\(callType)")
        CallEngine.shared.reportIncomingCall(
            callId: callId,
            callType: callType,
            displayName: targetName,
            pictureUrl: nil,
            metadata: metadata,
            completion: nil
        )
    }

    public func startCall(
        callId: String,
        callType: String,
        targetName: String,
        metadata: String?
    ) throws {
        logger.info("🎯 startCall ▶ js → native: \(callId), type=\(callType)")
        CallEngine.shared.startCall(
            callId: callId,
            callType: callType,
            targetName: targetName,
            metadata: metadata
        )
    }

    /// This function is called from JS to signal that a remote party has answered an outgoing call.
    public func callAnswered(callId: String) throws {
        logger.info("🎯 callAnswered ▶ js → native: \(callId)")
        CallEngine.shared.connectOutgoingCall(callId: callId)
    }

    public func setOnHold(callId: String, onHold: Bool) throws {
        logger.info("🎯 setOnHold ▶ js → native: \(callId), onHold=\(onHold)")
        CallEngine.shared.setOnHold(callId: callId, onHold: onHold)
    }

    public func setMuted(callId: String, muted: Bool) throws {
        logger.info("🎯 setMuted ▶ js → native: \(callId), muted=\(muted)")
        CallEngine.shared.setMuted(callId: callId, muted: muted)
    }

    public func updateDisplayCallInformation(callId: String, callerName: String) throws {
        logger.info("🎯 updateDisplayCallInfo ▶ js → native: \(callId), name=\(callerName)")
        CallEngine.shared.updateDisplayCallInformation(
            callId: callId,
            callerName: callerName
        )
    }

    public func registerVoIPTokenListener(listener: @escaping (String) -> Void)
        throws -> () -> Void
    {
        logger.info("🎯 registerVoIPTokenListener ▶ js → native")
        VoIPTokenManager.shared.registerTokenListener { token in
            self.logger.info("🎯 voip token received")
            listener(token)
        }
        return {
            self.logger.info("🎯 unregisterVoIPTokenListener ▶ js → native")
            VoIPTokenManager.shared.unregisterTokenListener()
        }
    }

    public func hasActiveCall() throws -> Bool {
        logger.info("🎯 hasActiveCall ▶ js → native")
        return CallEngine.shared.hasActiveCalls()
    }

    public func requestOverlayPermissionAndroid() throws -> Bool {
        // This is an Android-specific method, so we provide a no-op success case for iOS.
        return true
    }

    public func hasOverlayPermissionAndroid() throws -> Bool {
        // This is an Android-specific method, so we provide a no-op success case for iOS.
        return true
    }
}
